contracts/libs/data-structures/IncrementalMerkleTree.sol

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+/**
+ * @title Incremental Merkle Tree module
+ *
+ * This implementation is a modification of the Incremental Merkle Tree data structure mentioned
+ * in [Deposit Contract Verification](https://github.com/runtimeverification/deposit-contract-verification/blob/master/deposit-contract-verification.pdf).
+ *
+ * This implementation aims to optimise and improve the original data structure.
+ *
+ * The main differences are:
+ * - No explicit constructor; the tree is initialised when the first element is added.
+ * - Growth is not constrained; the height of the tree grows as elements are added
+ *
+ * Zero hashes are computed each time the getRoot function is called.
+ *
+ * You can find the gas analysis [here](https://colab.research.google.com/drive/1I-nGhb_LdcWW4pvQtE2F9JVDv371rO5f?usp=sharing).
```

Comment:
Let's not include personal external links for now. We will have a DLSL site in the future. You can include the gas usage right here